### PR TITLE
[Vaults] Restrict access to conversations and messages

### LINKED ISSUE #6903: [Vaults UI] Restrict access to a conversation based on access to the agent configurations used
Requires https://github.com/dust-tt/dust/issues/6763 to be done.

### PR DESCRIPTION
**Pull request by philipperolet**
## Description

Fixes https://github.com/dust-tt/dust/issues/6903

Left a few comments on the PR for reviewers to check.

Now that agents have ACLs, conversations and messages should be restricted to users that have access to the related assistants.

Restriction is implemented on conversations by: 
- checking if any of its agent messages shouldn't be accessed by the user. ACLs are naturally present on agent messages via their attached agent configuration's groupIds.
- in cases where we don't have access to all the messages of a conversation, checking conversation access via a new function `canAccessConversation`,  relying on conversation participants---that will now return an error if one of the agents or more is not accessible to the user;

Summary: 
- checked all references to AgentMessage and Message models (details below) and acted where needed
- changed getConversation to a Result with ConversationPermissionError / ConversationNotFoundError
  - wrapped conversation api errors in a dedicated function, drying quite a bit
- did the same with getConversationWithoutContent, further drying
- double-check: all endpoints, and references to Conversation

### Latency
Some calls become more heavy due to the addition of a `canAccessConversation` call relying on `fetchConversationParticipants` and as such making 2 additional database queries. The calls are indexed and query results are ~small volume, but still worth monitoring. This is the cost of checking per-conversation in addition to per-messages.

### Testing
Test scenarios:
- with allowed agents, all below should work
  - UX 
	- start a convo :heavy_check_mark: 
	- read a convo :heavy_check_mark: 
	- post a message to a convo :heavy_check_mark: 
	- destroy a convo :heavy_check_mark: 
  - API v1
	- start a convo, 
	- read a convo, 
	- post a message to a convo, 
	- destroy a convo
- with unallowed agents, all below should fail
  - start a convo (only API v1)
  - read a convo (both UX and v1) :heavy_check_mark: 
  - post a message to a convo (only v1) 
  - destroy a convo (only v1)

## Deploy Plan
test on front-edge
front

### CODE REVIEW

File: front/lib/api/assistant/conversation/helper.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import {
```
Comment (philipperolet) [inline front/lib/api/assistant/conversation/helper.ts:1]: Hesitated with a wrapper here; it's used in ~all api routes with conversations, I ~think the direct function call is more readable although a bit less verbose, open to discussion

File: front/lib/api/assistant/conversation.ts [hunk 2]
```diff
@@ -269,7 +279,7 @@ export async function getConversation(
   auth: Authenticator,
   conversationId: string,
   includeDeleted?: boolean
-): Promise<ConversationType | null> {
+): Promise<Result<ConversationType, ConversationPermissionError>> {
```
Comment (philipperolet) [inline front/lib/api/assistant/conversation.ts:282]: Note: did *not* do the restriction check for the `getUserConversations` function a few lines above here (above `getConversation`)

Reason is it would become too heavy. Not checking is fine IMO since 1. already filtered by user being a participant, 2. not leaking conversation content. If conversation content on a conversation that became private were to be requested, then the restriction *would* apply.

File: front/lib/api/assistant/conversation.ts [hunk 3]
```diff
@@ -693,16 +701,29 @@ export async function* postUserMessage(
...
-      )
+    await Promise.all(
+      mentions.filter(isAgentMention).map((mention) => {
+        return getLightAgentConfiguration(auth, mention.configurationId);
+      })
     ),
     await createOrUpdateParticipation(),
   ]);
-  const agentConfigurations = results[0];
+
```
Comment (philipperolet) [inline front/lib/api/assistant/conversation.ts:711]: Note: in `edit/postUserMessage`, agents to which the user is not authorized are already filtered out. 

Still, we check and fail if one of the agents was not accessible, since it seems better to be explicit about it to the user than silently ignore the mention. 

The other option would be to use the new `canAccessConversation` fn relying on conversation participants. It would be somewhat cleaner, but makes the calls to `postUserMessage` more heavy, and since it's one of the most used and critical part of our app the choice I made here is to keep it lighter; happy to discuss

File: front/lib/api/assistant/participants.ts [hunk 4]
```diff
@@ -109,6 +115,12 @@ export async function fetchConversationParticipants(
     fetchAllAgentsById(auth, [...agentConfigurationIds]),
   ]);
 
+  // if less agents than agentConfigurationIds, it means some agents are forbidden
```
Comment (philipperolet) [inline front/lib/api/assistant/participants.ts:118]: Here the check on length is used to determine whether some agents were forbidden. 

This is a compromise, since it's not the cleanest way to do this check. The alternative is refactoring all getAgentConfiguration(s) calls (~30 of them + potentially the parent callers) to return a Result type rather than directly the agent configuration. It's a cleaner option, but heavier--not only to do now but in code every time we'll call getAgentConfig(s), we'll have to unpack / handle which is not always useful.

Hesitating a bit there so happy to get the reviewers' opinions. Not opposed to go for the cleaner but heavier way.

File: front/lib/api/assistant/conversation.ts [hunk 5]
```diff
@@ -677,16 +701,29 @@ export async function* postUserMessage(
...
+  // casting non-null is ok here because we check right below that there's no
+  // nulls and err if there is
+  const agentConfigurations = results[0] as LightAgentConfigurationType[];
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:714]: Not sure to understand why we need this 🤔. The code casts `results[0]` to `LightAgentConfigurationType[]` assuming it's non-null, but then checks for nulls afterward 🤯. Instead of casting immediately, consider handling potential null values first. This would improve clarity and prevent assumptions.

File: front/lib/api/assistant/conversation.ts [hunk 6]
```diff
@@ -677,16 +701,29 @@ export async function* postUserMessage(
   }
 
   const results = await Promise.all([
-    removeNulls(
-      await Promise.all(
-        mentions.filter(isAgentMention).map((mention) => {
-          return getLightAgentConfiguration(auth, mention.configurationId);
-        })
-      )
+    await Promise.all(
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:704]: The `await` here breaks the || of the `Promise.all()`, right? IIRC `Promise.all()` should take promises, here we await them.

File: front/lib/api/assistant/conversation.ts [hunk 7]
```diff
@@ -1132,18 +1170,29 @@ export async function* editUserMessage(
...
     ),
-
-    createOrUpdateParticipation(),
+    await createOrUpdateParticipation(),
   ]);
 
-  const agentConfigurations = results[0];
+  // casting non-null is ok here because we check right below that there's no
+  // nulls and err if there is
+  const agentConfigurations = results[0] as LightAgentConfigurationType[];
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:1183]: Same here.

File: front/lib/api/assistant/conversation.ts [hunk 8]
```diff
@@ -1470,6 +1519,11 @@ export async function* retryAgentMessage(
   void
 > {
   class AgentMessageError extends Error {}
+
+  if (!canReadMessage(auth, message)) {
+    throw new AgentMessageError("Agent cannot be used by this user");
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:1524]: IMHO, we should not throw for permissions reasons around conversation since this resource can be easily shared, we should expect this behavior to happen.

File: types/src/front/assistant/conversation.ts [hunk 9]
```diff
@@ -219,3 +219,15 @@ export interface ConversationParticipantsType {
   agents: AgentParticipant[];
   users: UserParticipant[];
 }
+
+export class ConversationPermissionError extends Error {
```
Comment (flvndvd) [inline types/src/front/assistant/conversation.ts:223]: Let's not introduce one error per case, we have a generic `DustError` that can be used for those two use cases.

**Review by flvndvd — COMMENTED** [review]
Fairly surprise that we need to update that many different places. They all seem to have in common the `Conversation.findOne()`. Any chance we could have a wrapper around this and check the permissions there?

Having to add the `canAccessConversation` check in 8+ different places does not seem very robust.

File: front/lib/api/assistant/conversation.ts [hunk 10]
```diff
@@ -143,18 +149,22 @@ export async function updateConversation(
...
     visibility: visibility,
   });
 
-  const c = await getConversation(auth, conversationId);
+  const cRes = await getConversation(auth, conversationId);
 
-  if (!c) {
-    throw new Error(`Conversation ${conversationId} not found`);
+  if (cRes.isErr()) {
+    throw cRes.error;
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:164]: We used to throw as this error was unexpected but now that this is a more versatile error do we really want to throw?

File: front/lib/api/assistant/conversation.ts [hunk 11]
```diff
@@ -143,18 +149,22 @@ export async function updateConversation(
     throw new Error(`Conversation ${conversationId} not found`);
   }
 
+  if (!(await canAccessConversation(auth, conversation))) {
+    throw new Error("Conversation access denied");
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:153]: Is this really a throw here? A throw is a panic.

File: front/lib/api/assistant/conversation.ts [hunk 12]
```diff
@@ -188,6 +198,10 @@ export async function deleteConversation(
...
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:202]: Same here

File: front/lib/api/assistant/conversation.ts [hunk 13]
```diff
@@ -365,17 +385,21 @@ export async function getConversationWithoutContent(
   });
 
   if (!conversation) {
-    return null;
+    return new Err(new ConversationNotFoundError());
   }
 
-  return {
+  if (!(await canAccessConversation(auth, conversation))) {
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:391]: This defeats the purpose of getConversationWithoutContent whose sole purpose is to read a conversation fast without loading messages.

Eg. in the praticipants endpoint we call it and then fetch the participants but now that call also fetch the participants which means we fetch the participants twcie

**Review by spolu — COMMENTED** [review]
Left a few comments, but thinking more about it I think we want to bite the bullet and bubble up groups on the conversation object.

The code to compute the new groups in postUserMessage would be 3 lines. We have a transaction ready to receive that mutate on the conversation object that we can do in parallel of creating the messages/agent_messages/user_messages/mentions

Why?

You introduce 2 ways here to decide if we can see a conversation to optimize. This is a weakness as 2 flows to enforce a critical security feature is a foot gun. Checking if a conversation is accessible should be trivial and ideally instantaneous so that we never ever hesitate checking it.

Downside:
- 1 more write at message creation but done concurrently (security benefits trump this IMHO)
- Risk of getting out of sync (~impossible if placed in transaction)

Upside:
- Trivial to use and non block `canAccessConversation` improving our safety


Given all that I feel like this is highly desirable to bubble up.

File: front/lib/api/assistant/conversation.ts (hunk 11 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:153]: Following the convention for that function, just above you can see a throw for conversation not found; in both cases it's really not gonna work if you try to update the conversation, and should basically never happen

We could refactor the function to return a result, but the PR being big I thought best to leave it oustide scope

File: front/lib/api/assistant/conversation.ts (hunk 10 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:164]: IMO, both cases are quite unexpected; update of a forbidden convo is impossible via the UI and adversarial via APi so I thought it was fine to throw

Not strong on this, so ok to evolve

File: front/lib/api/assistant/conversation.ts (hunk 12 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:202]: same answer

File: front/lib/api/assistant/conversation.ts (hunk 13 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:391]: Yes, this was a conscious tradeoff as described in the PR risks. I think it's the least damageable option

I went through all the callers of `getConversationWithoutContent` and the permission check was legit every time, except for the participants endpoint. So we could move the check out but IMO would make the code more brittle because current stats show the check must be done almost all the time (1 exception out of ~15 IIRC)

Basically we're putting mental load on the callers and the check will likely be missed at some point. Also, having a Result here seemed more robust and allowed some code drying

File: front/lib/api/assistant/conversation.ts (hunk 6 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:704]: :+1: well spotted, we did had the issue before too I guess ; I refactored without noticing

File: front/lib/api/assistant/conversation.ts (hunk 5 above)
Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:714]: Right, thought the comment would clarify but true it's better below :+1:

**Comment by philipperolet** [comment]
> Left a few comments, but thinking more about it I think we want to bite the bullet and bubble up groups on the conversation object.

Good points, but would like to discuss. I had been unclear and this helped me clear my mind, see below.

> 
> The code to compute the new groups in postUserMessage would be 3 lines. We have a transaction ready to receive that mutate on the conversation object that we can do in parallel of creating the messages/agent_messages/user_messages/mentions
> 
> Why?
> 
> You introduce 2 ways here to decide if we can see a conversation to optimize. This is a weakness as 2 flows to enforce a critical security feature is a foot gun. 

2 flows for the same purpose would be an issue, especially if it's solely to optimize, I agree. But here there are two separate, arguably unrelated flows:
- checking for access on messages;
- checking for access on converations.

What I wrote in my PR description was unclear and misguided, leaving to believe that messages check was a way of optimizing conversation check. I apologize for the lack of clarity. I think we should in any case do both checks, independently of each others.

In particular, we should enforce auth on messages, via message-related functions that take `auth` as arg. Removing this would be a footgun too; future us will reuse those functions, ACLs on messages are very clearly defined, more than on conversations.

Regarding the fact that one flow is used to optimize the other:
The only situation in which messages are used to optimize conversation check is on `getConversation`. It could be avoided and we could use the regular `canAccessConversation` to do that instead, to completely separate the two flows. It would not be an issue because `getConversation` is already heavy since it grabs all content.

If we did this the flows would become completely separate. I reckon I did the optimization because it was here right under my nose, but it's not really needed and i'm fine to discard it for the purpose of clarity.

> Checking if a conversation is accessible should be trivial 

I think it is already trivial, it's calling a function named as described

> and ideally instantaneous so that we never ever hesitate checking it.

I reckon it's a good point; it is true that here we add a db read when grabbing a conversation. But the converse, adding a DB write when we post a message, does not seem better to me since more messages are posted than conversations grabbed. IMO it's not that much of a heavy call, but let's say it was: if  parallelization could be done too on the db read, e.g. with @flvndvd proposal of wrapping, would that help the current solution? 

In ~all actual cases, we did have to check it so we wouldn't hesitate anyway. The call is not that heavy.
 
> Downside:
> 
> * 1 more write at message creation but done concurrently (security benefits trump this IMHO)

> * Risk of getting out of sync (~impossible if placed in transaction)

Personally scared of the out of sync risk, the transaction locally prevents some issues but in our wide complex system other problems can arise; needing sync can result in compounded complexity hard to anticipate.
 
> Upside:
> 
> * Trivial to use and non block `canAccessConversation` improving our safety

IMO canAccessConversation is already trivial to use. I agree the blocking part is a tradeoff that we can discuss

> 
> Given all that I feel like this is highly desirable to bubble up.

Let's have a quick final chat IRL to decide

**Comment by philipperolet** [comment]
> Fairly surprise that we need to update that many different places. They all seem to have in common the `Conversation.findOne()`. Any chance we could have a wrapper around this and check the permissions there?
> 
> Having to add the `canAccessConversation` check in 8+ different places does not seem very robust.

Interesting, TBD IRL along with @spolu 's points

**Comment by spolu** [comment]
>> Checking if a conversation is accessible should be trivial

> I think it is already trivial, it's calling a function named as described

Sorry by trivial I meant costless so that there is no friction.

> getConversation is already heavy since it grabs all content.

We thrive to make it lighter not heavier the fact that it's already heavy is not a good reason IMHO to make it heavier :)

> In ~all actual cases, we did have to check it so we wouldn't hesitate anyway. The call is not that heavy.

In the current implementation I would hesitate using it give the added cost of the call.

> Personally scared of the out of sync risk, the transaction locally prevents some issues but in our wide complex system other problems can arise; needing sync can result in compounded complexity hard to anticipate.

Can you unpack your FUD because this is just FUD here. The only place that impacts the groups is the addition of a UserMessage no?

On the two flows:

It's not clear to me that we want to build right away a message level ACL. This might be desirable defense in depth in the future but right now everything flows through a conversation (messages are not shareable etc..). Less code is more with security IMHO.

**Comment by philipperolet** [comment]
> 
> We thrive to make it lighter not heavier the fact that it's already heavy is not a good reason IMHO to make it heavier :)

If it was a lot heavier I'd agree. Here, it's marginally heavier only VS the size of the call, fetching the whole convo. And when we want a light call there's `getConvoWOContent`

> > In ~all actual cases, we did have to check it so we wouldn't hesitate anyway. The call is not that heavy.
> 
> In the current implementation I would hesitate using it give the added cost of the call.

We can make the call lighter an parallelize. The DB write in postUserMessage has a cost too, and is more frequent than the conversation check

> 
> > Personally scared of the out of sync risk, the transaction locally prevents some issues but in our wide complex system other problems can arise; needing sync can result in compounded complexity hard to anticipate.
> 

> Can you unpack your FUD because this is just FUD here. The only place that impacts the groups is the addition of a UserMessage no?
> 
It is indeed admittedly fud: compounded complexity hard to anticipate so tough to unpack; if you disagree on the fact that this kind of sync is bad long term for our codebase, I trust your judgment since mine is at this time only intuition

> On the two flows:
> 
> It's not clear to me that we want to build right away a message level ACL. This might be desirable defense in depth in the future but right now everything flows through a conversation (messages are not shareable etc..). Less code is more with security IMHO.

Let's discuss, IMO it's a simple  costless natural check, and avoids a footgun for every coder that will decide to use the `xxxMessage(auth,yyy)` function

**Comment by spolu** [comment]
Let's continue the discussion IRL tomorrow morning :+1: :+1:

**Comment by philipperolet** [comment]
Related decision: https://github.com/dust-tt/decisions/issues/250
As per IRL discussion at 16:30, we move towards denormalization : conversations will get a `groupIds`
PR to be adapted accordingly

**Comment by philipperolet** [comment]
Closing, adapted PR with follow-up work is on  #7746
